cnap/userv/websocket_server.py

```diff
@@ -0,0 +1,322 @@
+publish the streams to an HTML5 based front-end SPA (Single Page Application).
+
+It is based on asyncio and coroutine programming models since most operations
+are IO bound.
+"""
+
+import os
+import sys
+import asyncio
+import logging
+import signal
+import http
+
+import redis.asyncio as redis
+import websockets.exceptions
+import websockets.server
+
+LOG = logging.getLogger(__name__)
+
+STREAM_WEBSOCKET_PORT = 31611
+PIPELINE_TABLE = "Pipeline-table"
+MAX_RECONNECTION_TIMES = 5
+
+class StreamWebSocketServer:
+    """
+    Stream WebSocket Server to publish streams via WebSocket from the broker.
+    """
+
+    def __init__(self):
+        self._stream_broker_redis_host = self._get_env(
+            "STREAM_BROKER_REDIS_HOST", "127.0.0.1")
+        self._stream_broker_redis_port = self._get_env(
+            "STREAM_BROKER_REDIS_PORT", "6379")
+        self._pipelines = {}
+        self._users = {}
+
+    @staticmethod
+    def _get_env(key, default=None):
+        """
+        Get environment variable.
+        """
+        if key not in os.environ:
+            LOG.warning("Could not find the key %s in the environment, "
+                        "use default value %s", key, str(default))
+            return default
+        return os.environ[key]
+
+    def _add_pipeline(self, sid: str):
+        """
+        Add pipeline.
+
+        Args:
+            sid: the id of pipeline to add.
+        """
+        LOG.info("add pipeline: %s", sid)
+        self._pipelines[sid] = asyncio.get_event_loop().create_task(
+            self._stream_publish_task(sid))
+
+    def _del_pipeline(self, sid: str):
+        """
+        Delete pipeline.
+
+        Args:
+            sid: the id of pipeline to delete.
+        """
+        LOG.info("remove pipeline: %s", sid)
+        self._pipelines[sid].cancel()
+        del self._pipelines[sid]
+
+    def _on_update(self, pipeline_list: list[str]):
+        """
+        Update pipelines.
+
+        Args:
+            pipeline_list: the list of pipeline ids to update.
+        """
+        for new_item in pipeline_list:
+            if new_item not in self._pipelines:
+                self._add_pipeline(new_item)
+
+        for old_item in list(self._pipelines.keys()):
+            if old_item not in pipeline_list:
+                self._del_pipeline(old_item)
+
+    async def _websocket_server_task(self, wsobj: websockets.server.WebSocketServerProtocol,
+                                     path: str):
+        """
+        Connection handler for websocket server connection.
+
+        Args:
+            wsobj: the WebSocketServerProtocol object for websocket server connection.
+            path: the path for websocket server connection.
+        """
+        LOG.info("WebSocket server task start: %s path: %s", str(wsobj), path)
+        target = path[1:]  # skip / prefix
+
+        if len(list(self._users.keys())) > 100:
```

Comment:
Make the number as a constant. `MAX_CLIENT_CONNECTIONS`

Reply:
Thanks, fixed.